### ISSUE TITLE
Disable L3HA by default

### ISSUE THREAD

**Issue by BjoernT** [rcbops/rpc-openstack#1149]
##### What version (commit) do you encounter this issue on?

r12.1.2

##### What type of issue is this (bug, feature, documentation, regression)?

Bug

##### What is the expected behavior?

L3HA working correctly

##### What is actually happening?

https://bugs.launchpad.net/neutron/+bug/1590845
Duplicate: https://bugs.launchpad.net/neutron/+bug/1591386

##### How can we reproduce this issue?

Spinning up multiple L3HA enabled router will result in nearly 100% with ports stuck in BUILD state
which may or may not pass traffic.

##### How can we verify your issue is addressed?

Once upstream bug is fixed

##### If this is a large issue, what can be done immediately to alleviate the issue while working on a longer-term solution?

Normal issue since it's just a config change 

##### What is your evaluation on the importance of this issue, and why?

It is prio 1 because neutron router created over Horizon can not be configured in a non L3HA mode

----------------------------------------------------------------------------------------------------------------------------
###### For each pull request, update the branch line to reference it.
- [x] master
- [ ] liberty-12.2

##### Upgrade impact. Consider the impact on moving between major, minor and patch releases.
- [ ] This change does not impact upgrades
- [ ] This change addresses a potential upgrade impact
- [ ] The upgrade impact is being tracked by another issue.

The issue tracking the upgrade impact is:

##### Testing impact
- [ ] An issue/pull-request has been create or exists to track the testing gap, and is referenced below
- [ ] This change includes additional testing
- [ ] This change is tested by the existing setup

Additional testing links:

##### Release note
- [x] This change includes a release note.
- [ ] This change does not require a release note.

##### Does your change require documentation?
- [ ] Yes, and I submitted an issue to [Docs issues](https://github.com/rackerlabs/docs-rpc/issues "Docs issues")
- [x] Yes, and I contributed a documentation patch to [docs-rpc](https://github.com/rackerlabs/docs-rpc "docs-rpc")
- [ ] No documentation required for this change

Additional documentation links (please add links if you have ticked yes):

**Comment by BjoernT** [comment]
FYI, beyond the reported issue, Neutron has several other issues open.
Hence the recommendation to turn it off before it becomes a liability in customer environment.

https://bugs.launchpad.net/neutron/+bug/1523780 Race between HA router create and HA router delete
https://bugs.launchpad.net/neutron/+bug/1533455 Stale processes lives after a fanout deleting HA router RPC between L3 agents
https://bugs.launchpad.net/neutron/+bug/1488015 Potential race condition in l3-ha handling of l2pop initial master selection
https://bugs.launchpad.net/neutron/+bug/1533443 ML2: cannot update HA router ha_port states after race between HA router creating and deleting
https://bugs.launchpad.net/neutron/+bug/1533440 Race between deleting last HA router and a new HA router API call

**Comment by git-harry** [comment]
@Apsu would you mind giving your opinion regarding turning off L3 HA.

**Comment by Apsu** [comment]
Ok, so after some research and digging around, my recommendation is to disable L3HA at this point in time *unless* there’s a customer or customers who ran into scaling issues without L3HA and has a specific need that it supposedly is solving. But given the bugs and poor behavior, I’d suspect that they’d rather deal with potential failover issues than L3HA issues, so my recommendation is still to disable it.

**Comment by git-harry** [comment]
There's no explanation for why this is blocked. I'm assuming it's because upstream issues are referenced. If you view the comments here and on the associated pull request the consensus was to disable L3HA and so work around the fact that there are bugs by not using this feature. This issue should be used to implement that change.

Note the existing pull request has been done by @BjoernT, you will need to work with him to get that completed. In practise this will likely mean making a new pull request because, as stated in our [contributing guidelines](https://github.com/rcbops/rpc-openstack/blob/master/CONTRIBUTING.md#selecting-an-issue-to-work-on), we are responsible for the work we commit to, not support, and so it is our responsibility to keep things moving forward.

**Comment by git-harry** [comment]
@alextricity25 based on you moving this to doing I'm assuming it's not blocked and so I'll remove the label.

**Comment by alextricity25** [comment]
For updates, please see the PR#1150

**Comment by alextricity25** [comment]
Due to https://review.openstack.org/#/c/347998/, simply turning off l3_ha via overrides won't work in mitaka(since the os_neutron role can't re-run). We will have to wait until RPCO includes this change before turning off l3_ha in mitaka.

For now, i'm going to see if I can get this change into liberty. I'll make a new PR into liberty-12.2 once I'm done with my research. There are many questions to be answered:
1. Will the IPs of a L3 neutron router auto-migrate from the "secondary" networking node to the "primary" after disabling l3ha? Or will they all just stay where they are at?
2. If so, that probably means some down time. Is there any network down time for these migrations to happen? 
3. Do the other l3_ha configurations need to be addressed, or will they be ignored when l3_ha is false? https://github.com/openstack/openstack-ansible/blob/liberty/playbooks/roles/os_neutron/templates/neutron.conf.j2#L60-L64
4. What about lingering l3 agents? If we turn off l3ha on an existing environment, do the extra l3 agents need to be addressed? I'm not so sure if the extra l3 agent is needed. I'll have to do more digging into the neutron-ha-tool and how it handles the linux network namespaces for neutron routers.


I should also note that https://bugs.launchpad.net/neutron/+bug/159084 has a fix released in upstream now, so I'm not sure if we want to put a card in the backlog for turning L3HA back on in future releases(Newton).

**Comment by git-harry** [comment]
@evrardjp @prometheanfire @Apsu @tomc603 would you please provide your thoughts regarding disabling l3_ha in Mitaka. This issue was originally raised to make the change for Liberty, is it still appropriate to do this in Mitaka?

**Comment by alextricity25** [comment]
Also,
5.We have to consider how lingering HA networks will be handled after L3HA has been turned off(after an upgrade, or even just by setting l3_ha to false and re-running the os_neutron role). 
6. How about lingering router namespaces? Since L3HA creates *at least*  min_l3_agents_per_router namespaces. How do we deal with the passive router namespaces? Will those be cleaned up, or will we have to do that?

I'll do some research on all this and make updates here.

**Comment by prometheanfire** [comment]
Unless https://bugs.launchpad.net/neutron/+bug/1590845 is backported to mitaka. we'll need to disable it.  It'd be nice to just get it backported at least to mitaka.  Also, it was my understanding that the arp entries are now (in mitaka) populated by flood and store, but this is more specific to L2Population.

**Comment by alextricity25** [comment]
I recorded a session of what the repercussions of turning off l3ha are:
https://www.youtube.com/watch?v=7ibwjZqQN9Q

That should address some of the questions above, but there is still more to be researched.
* lingering keepalived pr

**Comment by alextricity25** [comment]
We could disable this in mitaka as well, but we have to wait for https://review.openstack.org/#/c/347998/ to be included in RPCO. Otherwise operators won't be able to re-run the neutron playbooks(unless they are doing a minor release). This is why the mitaka PR for this is blocked:
https://github.com/rcbops/rpc-openstack/pull/1150

**Comment by prometheanfire** [comment]
ya, saw that we'd need to get that review :P

**Comment by git-harry** [comment]
@alextricity25 the first release of Mitaka will not include support for major upgrades.

**Comment by alextricity25** [comment]
@git-harry That's great. That means having L3HA set to false by default for liberty/mitaka greenfield deployments is fine. But what about minor upgrades for liberty builds?

**Comment by alextricity25** [comment]
I've made a PR to liberty #1284 . However, there are a few things support needs to be aware of before turning off l3_ha *on an existing environment with l3_ha already on*. 

1. Will the IPs of a L3 neutron router auto-migrate from the "secondary" networking node to the "primary" after disabling l3ha? Or will they all just stay where they are?
They just stay in whatever networking host they are on.

2. If so, that probably means some down time. Is there any network down time for these migrations to happen? 
No down time, since the IPs stay on their network namespace after l3_ha has been set to False and the os_neutron role has re-run. 

3. Do the other l3_ha configurations need to be addressed, or will they be ignored when l3_ha is false? https://github.com/openstack/openstack-ansible/blob/liberty/playbooks/roles/os_neutron/templates/neutron.conf.j2#L60-L64
They will be ignored.

4.We have to consider how lingering HA networks will be handled after L3HA has been turned off(after an upgrade, or even just by setting l3_ha to false and re-running the os_neutron role).
The HA networks stick around even after turning l3_ha off. They will probably have to be deleted manually. I tried deleting the HA ports for a router manually. I also tried deleting the HA network. Although the router stuck around, the router namespace in the neutron agent container disappeared after restarting the l3_agent. There is probably a special procedure that needs to be followed here if you want to get rid of the HA networks. Also, keep in mind that re-creating the routers and re-attaching the subnets is also possible, but there is definitely some downtime for the VMs attached to said subnet.

5. How about lingering router namespaces? Since L3HA creates at least min_l3_agents_per_router namespaces. How do we deal with the passive router namespaces? Will those be cleaned up, or will we have to do that?
These namspaces actually disappeared once I deleted the HA ports and HA networks for an associated routers, causing downtime for me. The fix for me was to re-create the routers after l3_ha had been turned off, and re-attach the subnets. In order to do this, the floating-ips on VMs associated with said subnet had to be unassociated. Which is not ideal.

In a nutshell, turning off L3HA on an existing system isn't something that can be done without having to perform other tasks to completely remove the HA routers. It's also risky, causing network downtime for any VMs on subnets attached to the HA router.

**Comment by mancdaz** [comment]
Thanks @alextricity25 for all your research and notes here. It seems like disabling it is the right thing to do, but we need to document the steps a deployer (support) should take when moving from an environment where this has already been enabled. 

This means that in addition to the PRs to disable it, you should  make a PR to the docs repo detailing all of this, what to look for, what the impact is, how to perform the 'migration' etc. This is likely something support will need to do in specific maintenances for those customers, or at least in a special section of a patch update maintenance.

**Comment by alextricity25** [comment]
@mancdaz Sounds good. I'll sync up with Robb to figure out a good place for this. Thanks!

**Comment by alextricity25** [comment]
The two PRs are ready for review:
https://github.com/rcbops/rpc-openstack/pull/1284 (liberty)
https://github.com/rcbops/rpc-openstack/pull/1285 (mitaka)

@git-harry Who did you say you wanted to take a look at these?

**Comment by git-harry** [comment]
@Apsu @tomc603 @evrardjp @prometheanfire please would you review the pull requests associated with this issue.

**Comment by alextricity25** [comment]
UPDATE:
I tried updating the router "HA" attribute **before** turning l3ha off in the neutron config, and that seemed to do the trick. The catch is that the router must be in admin_state DOWN before it allows you to update the HA attribute, but at least this way you won't have to delete the router and re-create it. I'll update the docs in docs-rpc

**Comment by tomc603** [comment]
For documentation purposes, this is going to be related to #973. By disabling l3ha we also need to enable l2pop. But, this means things like Nova live migration won't work, and we'll have to deal with the inherent complexity of l2pop.

**Comment by prometheanfire** [comment]
it's my understanding that l2pop can now do flood and learn on ml2/vxlan so it's slightly simpler than it used to be.

**Comment by BjoernT** [comment]
OK we are discussing the same thing between #973 and here, let's just pick one issue to continue discussing it

**Comment by tomc603** [comment]
Alright, there's some confusion here. I've linked the two issues in both directions for the purpose of future clarity. Almost six months ago, OSA committed changes that used l3ha by default and changed the way l2 population was done.

At some point in the future, we will no doubt enable l3ha. Having all of the relevant historical information easy to find is going to save us time and the confusion we're experiencing now.

**Comment by alextricity25** [comment]
Marking this as bocked because of what I mentioned on the PR:
https://github.com/rcbops/rpc-openstack/pull/1284

**Comment by git-harry** [comment]
@alextricity25 the issue doesn't need to be blocked just because the review process has identified the pull request requires changing.

**Comment by BjoernT** [comment]
Adding a new bug reference that L3HA is still not considered production ready: https://bugs.launchpad.net/neutron/+bug/1648242

**Comment by BjoernT** [comment]
@claco  FYI, the Neutron Team is testing various scenarios around the L3 agent/HA https://ethercalc.openstack.org/Pike-Neutron-L3-HA

**Comment by jamiramontez** [comment]
@alextricity25, what was the rationale for closing this ticket? Thanks!

**Comment by alextricity25** [comment]
The intent of this ticket was to "Disable L3HA by default". That has been done and documented, so I closed off the ticket.
The setting to disable l3ha is here:
https://github.com/rcbops/rpc-openstack/blob/newton-14.0/rpcd/etc/openstack_deploy/user_osa_variables_defaults.yml#L125-L129

**Comment by jamiramontez** [comment]
Thanks @alextricity25!
